[Feature] Add obstruction indicator

Estecka opened the pull request:
This makes some changes to the Static crosshair, so that a secondary "dynamic" crosshair will become visible if there is an obstacle between the character and the static crosshair.
Mechanically, the aiming is still completely identical to the Static crosshair. The difference is only visual:

https://youtu.be/cTxliMQeqTY

This is based on what I remember of the crosshair in Metal Gear Solid V. I find this much easier to aim than the current Adaptative/Dynamic Option. In the end that extra dynamic crosshair isn't really meant to be aimed, it's more of a feedback to tell the player they are not aiming where they intend to.

This is still a WIP, as there's probably some config options to be added around that.

Estecka commented on common/src/main/java/com/github/exopandora/shouldersurfing/mixins/MixinGui.java:
```diff
@@ -45,4 +52,29 @@ private boolean isFirstPerson(CameraType cameraType)
+		crosshairRenderer.preRenderCrosshair(guiGraphics.pose(), this.minecraft.getWindow());
+		if(!crosshairRenderer.doRenderCrosshair())
+		{
+			return;
```
I wanted to put the `return` at the very top instead, but kept it after the `preRender` to mimic how the old mixins worked. It there any reason you did it this way ?

Exopandora replied:
This is an relic of old times. In earlier versions, the dynamic crosshair position was interpolated (updated on client tick). For that reason it was required to update the position even when the dynamic crosshair was not active. But since this is no longer the case, feel free to change the order. :)

Exopandora commented on api/src/main/java/com/github/exopandora/shouldersurfing/api/model/HybridPickContext.java:
```diff
@@ -0,0 +1,25 @@
+import net.minecraft.world.level.ClipContext;
+import net.minecraft.world.phys.Vec3;
+
+public final class HybridPickContext
```
I don't think this class is necessary. All you want to show is the intersection point of the dynamic trace while the static crosshair is active. For that, you could reuse the dynamic trace. The "endPos" should be identical to the "endPos" of the static crosshair when aiming, as the player's rotation will be adjusted.

Estecka replied:
I tried reusing the DynamicPickContext at first, but encountered multiple issues with it:
- It takes a few frames for the character's rotation to fully snap to the main crosshair, so the secondary crosshair may become visible and lag slightly behind the main one whenever you move the camera, even if there is no obstacle in the way. This is strongly exacerbated by having the camera decoupled from the character's rotation (yes, even when aiming). There's no such lag on the *actual* dynamic crosshair, so I have no idea why it happens here.
- The secondary crosshair's position is completely wrong if the Pick Vector is set to "Camera" in the config (which is the default value).
- The secondary crosshair's position is also wrong when the character is not aiming, even if the camera is not decoupled, and even if the Pick Vector is set to "Player".

Commit ee06267 still uses the original `updateDynamicRaytrace` if you want to try it out. (Fabric only)

At the very least I would still need some way to override the context's Pick Vector, but overriding the endPos feels safer and more straightforward; it's a surefire way to avoid all of the weird edge cases like those, and the ones we might not even think about. In hindsight, I think I should also use the custom endPos to reduce the interactionRangeOverride, just to make sure the second crosshair doesn't accidentally shoot past the main one...

Exopandora commented on common/src/main/java/com/github/exopandora/shouldersurfing/client/CrosshairRenderer.java:
```diff
@@ -61,16 +69,44 @@ public boolean doRenderCrosshair()
 			(this.projected != null || !this.isCrosshairDynamic(Minecraft.getInstance().getCameraEntity()));
 	}
 	
+	public boolean doRenderSecondaryCrosshair()
+	{
+		return this.projected != null
+			&& this.instance.isShoulderSurfing()
+			&& Config.CLIENT.getCrosshairVisibility(Perspective.current()).doRender(Minecraft.getInstance().hitResult, this.instance.isAiming())
+			&& !this.isCrosshairDynamic(Minecraft.getInstance().getCameraEntity())
+			;
+	}
```
I think the secondary crosshair should only be shown when aiming. You can call `ShoulderSurfingImpl.shouldEntityAimAtTarget(LivingEntity, Minecraft)` for that.

Estecka replied:
I noticed there was a `IShoulderSurfing.isAiming()`, wouldn't this one be simpler, or does it do something different ?

Exopandora replied:
The naming is not ideal here, I have to admit. `IShoulderSurfing.isAiming()` only returns true when a the player is holding or using an adaptive item. It does not return `true` when the player is right clicking, attacking or pick blocking. All of that is handled in `ShoulderSurfingImpl.shouldEntityAimAtTarget(LivingEntity, Minecraft)`.

Estecka replied:
> It does not return true when the player is right clicking, attacking or pick blocking.

That's even better! If the player uses the camera as pick origin, showing obstructions for these actions is irrelevant. And if the player uses the character as the pick origin, they'll want to see the obstruction _before_ they try to perform the action, in which case they need to have the crosshair enabled at all time, not just when aiming.

Speaking of which, it looks like the pick origin options don't work properly:

![2024-10-21_19 46 05](https://github.com/user-attachments/assets/6ee4179f-4a3e-42d3-b571-08e051b16044)

Entity Pick Origin seems to work backward. In order to interact with the armour stand, I need to set the origin to "player", when I would actually expect that option to _prevent_ me from interacting with objects behind a wall.

Block Pick Origin doesn't seem to do anything. It always pick the blocks behind the wall, but can never pick the wall itself, regardless of what option I choose.

Exopandora replied:
There seems indeed to be a bug with the pick origin. It looks like it is related to the interaction distance. I will look into it. Thanks for reporting!

Exopandora replied:
The pick origin bug should be fixed now (this time I also pushed it). I also want to give some more context: When you set the pick origin to "PLAYER" the origin of the raytrace is not the players head. Instead, the position is computed as follows (abstract):
`camera.position + camera.lookVector * cameraZOffset`
Maybe this figure makes it easier to understand:

![pick_origin](https://github.com/user-attachments/assets/8b19e13c-03d2-44f9-8c68-0348bf4911cf)

The only difference between pick origin "PLAYER" and "CAMERA" is that camera allows you to target objects between the camera and the player, or in other words, it allows you to target objects behind the player. The director of the pick vector stays the same.

Estecka replied:
Thanks for the clarification, I did have that completely wrong here. ^^"
Good to know that raycasts from the the player itself are not actually possible.

In any case, as expected, `shouldEntityAimAtTarget` causes the obstruction crosshair to flicker in and out when doing menial actions, so I've gone for `isAiming` which is closer to how the Adaptative crosshair works. I should be ready to push that and the other config options soon.

Exopandora commented on common/src/main/java/com/github/exopandora/shouldersurfing/client/CrosshairRenderer.java:
```diff
@@ -61,16 +69,44 @@ public boolean doRenderCrosshair()
+			boolean isDynamic = this.isCrosshairDynamic(Minecraft.getInstance().getCameraEntity());
 			double interactionRangeOverride = Config.CLIENT.useCustomRaytraceDistance() ? Config.CLIENT.getCustomRaytraceDistance() : 0;
 			Player player = Minecraft.getInstance().player;
-			PickContext pickContext = new PickContext.Builder(camera).dynamicTrace().build();
+			PickContext.Builder pickContextBuilder = new PickContext.Builder(camera);
+			if (isDynamic)
+				pickContextBuilder.dynamicTrace();
+			PickContext pickContext = pickContextBuilder.build();
 			HitResult hitResult = this.instance.getObjectPicker().pick(pickContext, interactionRangeOverride, partialTick, player);
-			Vec3 position = hitResult.getLocation().subtract(camera.getPosition());
-			this.projected = project2D(position, modelViewMatrix, projectionMatrix);
+			Vec3 position = hitResult.getLocation();
+			Vec2f primaryCross = project2D(position.subtract(camera.getPosition()), modelViewMatrix, projectionMatrix);
+
+			if (isDynamic)
+				this.projected = primaryCross;
+			else {
+				pickContext = new PickContext.Builder(camera).hybridTrace(position).build();
+				hitResult = this.instance.getObjectPicker().pick(pickContext, interactionRangeOverride, partialTick, player);
+				position = hitResult.getLocation();
+				Vec2f secondaryCross = project2D(position.subtract(camera.getPosition()), modelViewMatrix, projectionMatrix);
+
+				// Don't render secondary if too close to primary
+				if (secondaryCross==null || secondaryCross.subtract(primaryCross).lengthSquared() < 16*16)
+					this.projected = null;
+				else
+					this.projected = secondaryCross;
+			}
```
I think this can be simplified by using only the dynamic trace (please correct me if I'm wrong):

```java
// this.projected has been renamed to this.crosshairOffset in the following code snippet

double interactionRangeOverride = Config.CLIENT.useCustomRaytraceDistance() ? Config.CLIENT.getCustomRaytraceDistance() : 0;
Player player = Minecraft.getInstance().player;
PickContext pickContext = new PickContext.Builder(camera).dynamicTrace().build();
HitResult hitResult = this.instance.getObjectPicker().pick(pickContext, interactionRangeOverride, partialTick, player);
Vec3 position = hitResult.getLocation().subtract(camera.getPosition());
Vec2f projected = project2D(position, modelViewMatrix, projectionMatrix);
Vec2f crosshairOffset = null;

// the following block was moved from preRenderCrosshair
if(projected != null)
{
	Window window = Minecraft.getInstance().getWindow();
	Vec2f screenSize = new Vec2f(window.getScreenWidth(), window.getScreenHeight());
	Vec2f center = screenSize.divide(2);
	crosshairOffset = projected.subtract(center).divide((float) window.getGuiScale());
}

if(crosshairOffset != null && (this.isCrosshairDynamic(Minecraft.getInstance().getCameraEntity()) || crosshairOffset.lengthSquared() > 256))
{
	this.crosshairOffset = crosshairOffset;
}
else
{
	this.crosshairOffset = null;
}
```

However, there might be an issue with the pick vector.

Estecka replied:
I think I'll still need the two PickContexts and raycasts (see the other comment), but moving some code over from `preRender` is a good idea. 👍 
With this, I can move all the checks for _whether_ the crosshair should be displayed into `doRenderSecondaryCrosshair`, instead of manually setting `crosshairOffset` to null.

Exopandora commented on common/src/main/java/com/github/exopandora/shouldersurfing/mixinducks/GuiDuck.java:
```diff
@@ -0,0 +1,9 @@
+
+public interface GuiDuck
+{
+	void shouldersurfing$RenderCrosshair(GuiGraphics guiGraphics, DeltaTracker deltaTracker);
```
The function name should continue with lower case after the $ sign.

Exopandora reviewed (commented):
Thanks for contributing! This looks like a really nice feature indeed. I've left you some comments on what could be improved. Here is also some general feedback:
- It would be nice to have the secondary crosshair only trigger when the player is aiming (already mentioned in the code comments). Maybe this should also be configurable.
- It would be nice to have a configurable distance limit for the secondary crosshair. Both onscreen (offset from main crosshair) and a configurable maximum distance from the players head to the intersection point in the world.
- It would be nice to have a general feature toggle for the secondary crosshair.
- Regarding code formatting, I could provide a .editorconfig for easy auto-formatting.
- Regarding naming, I would make the suggestion for "collision indicator" or "collision crosshair". Feel free to suggest another name!

Estecka commented:
Thanks for the feedback! I was planning on adding most of these options, I think all of these can be implemented easily.
About disabling the crosshair when not aiming think this is better left as an option than force-enabled.

I'll take the .editorconfig, I'm not sure that VsCode can handle them, but I'll see what I can do.

What about "obstruction indicator" for the name ?

Unrelated to anything, I had another idea of changing the shape of the main crosshair when the second one is visible, making it either an X or an incomplete square. That should help make the second crosshair's appearance and disappearance more noticeable. That may require a bit more mixinery though, so I'll only look into this after the config options are done.

Exopandora commented:
> What about "obstruction indicator" for the name ?

That would also work. :)

> Unrelated to anything, I had another idea of changing the shape of the main crosshair when the second one is visible, making it either an X or an incomplete square. That should help make the second crosshair's appearance and disappearance more noticeable. That may require a bit more mixinery though, so I'll only look into this after the config options are done.

This may (or may not) lead to compatibility problems with crosshair mods. But I'm curious to see your implementation.

Alternatively, we could also implement a similar indicator as PUBG has for it's grenade throwing mechanic. They show a red circle at the impact position in the world. Maybe we could render a small (transparent) red box in the world (maybe even with customizable colors?). But that would be of course a lot work. Here is a short video demonstation that I found on youtube: https://youtu.be/Mn3uR4bzfSk?feature=shared&t=217

Exopandora commented:
Almost forgot it, I added a .editorconfig to master.

Estecka commented:
For changing the shape of the crosshair, the simplest way would be a `@ModifyArg` around the sprite's identifier; most of the time it will pass through whatever was given as input. The worst another mod could do is take priority and use its own crosshair sprite instead of ours.

Alternatively, we could have a custom implementation of `renderCrosshair` used only for when we want to draw the custom sprite, and use the vanilla `renderCrosshair` the rest of the time. This way the effects of other mods would be limited to whichever crosshair is currently using the default sprite. This would also prevent other vanilla things like the F3 crosshair and the melee swing cooldown from being displayed in two separate places.


PUBG's throwing trajectory looks like a completely separate feature, that would deserve its own branch and PR. Figuring out the different trajectories for different weapons and projectiles sounds like the hard part, especially if modded items are involved; I don't feel like dabbling in ballistic.

Did you push the editorconfig yet ? Master seems unchanged right now.

Exopandora commented:
> PUBG's throwing trajectory looks like a completely separate feature, that would deserve its own branch and PR. Figuring out the different trajectories for different weapons and projectiles sounds like the hard part, especially if modded items are involved; I don't feel like dabbling in ballistic.

Sorry, I didn't mean to render the trajectory. Just the point of impact (the red circle). We already have the position when calculating the the trace for the secondary crosshair. But instead of calling project2D, we just use the `hitResult.getLocation()` for rendering a red box.

> Did you push the editorconfig yet ? Master seems unchanged right now.

My bad! Forgot to actually push the changes to GH. It's there now.

Estecka commented:
> Sorry, I didn't mean to render the trajectory. Just the point of impact (the red circle).

Oh I see what you mean now! Maybe it could be there as a style option, but I'm still partial to having an actual crosshair on the obstacle. I think the red marker's meaning is more ambiguous, some people might see it and think it does take gravity into account.

Here are the config options. Feel free to rename them or change the translations as you see fit. I wasn't sure what to put as the default for the maximum range, I set it to 20, which is roughly the range of an arrow fired horizontally on flat ground.
Unfortunately it doesn't look like VsCode can work with `ij_` styled editorconfig, but I still tried to rework my code to match your style.

Estecka commented:
Here's what it would look like when changing the crosshair's sprite. I've tried this with the mod Dynamic Crosshairs and everything worked as expected: https://youtu.be/5JvCZvq25A8 [[code diff](https://github.com/Estecka/mc-ShoulderSurfing/commit/958216d34ff54950d6d15dac16487e4e296dd082)]

There are two additional crosshair shapes, one replaces the static crosshair while aiming, and the other replaces the obstruction crosshair when not aiming (if enabled).

Exopandora commented:
Thank you for all the feedback!

> Here's what it would look like when changing the crosshair's sprite. I've tried this with the mod Dynamic Crosshairs and everything worked as expected: https://youtu.be/5JvCZvq25A8 [[code diff](https://github.com/Estecka/mc-ShoulderSurfing/commit/958216d34ff54950d6d15dac16487e4e296dd082)]

Unfortunately, the video private, so I could not watch it. But I did a test run with the commit and really liked the feature, so I merged it into this PR. I also updated the branch to the latest master commit (1.21.3) and adjusted the formatting and config phrasing.

What I noticed is that the "obstructed crosshair" renders as inverted:
![image](https://github.com/user-attachments/assets/bf2d71b7-94f0-4abd-9f4b-5dc1a977fce4)


I think it would be nicer if the center cross would render as normal, so it would always be red.

Otherwise, I think this PR is almost ready to be merged. Thanks for the hard work!

Estecka commented:
Oops, I _did_ set the video to unlisted while uploading, but didn't click some "next" button that only shows up when upload's done, so youtube still treated it as a draft...

> What I noticed is that the "obstructed crosshair" renders as inverted. I think it would be nicer if the center cross would render as normal, so it would always be red.

That would be possible, I'm a bit concerned about whether that red cross would still be noticeable on red backgrounds, like in the nether, but we'll just have to test and see what works best. Maybe we could "cheat" and make a higher-res texture that looks pixelated, but actually has a thin outline around the cross to make it stand out against all backgounds.

Estecka commented:
It's not as bad as I thought, the nether is still much darker than the crosshair. Turns out the worst place you could be is actually the badlands:
![Crosshair](https://github.com/user-attachments/assets/30000888-ecec-4d64-8751-41c8b845f41f)

In the end the player will be looking straight at that crosshair most of the time, so it's probably not a big issue anyways.
I'll do that then mark the PR as ready.

Estecka commented:
Done! A heads up if you're looking to backport:

With 1.21.3, all the RenderSystem calls that used to surround `blitSprite` in `renderCrosshair` are now included in the RenderType passed to it as parameter, so I removed them here.

For the red X, I switched the render type to `RenderType::guiTextureOverlay`, which in 1.21.1 would be equivalent to changing the parameters of `blendFuncSeparate`:
```java
RenderSystem.blendFuncSeparate(
    SourceFactor.SRC_ALPHA,
    DestFactor.ONE_MINUS_SRC_ALPHA,
    SourceFactor.ONE,
    DestFactor.ONE_MINUS_SRC_ALPHA
);

Exopandora commented:
> With 1.21.3, all the RenderSystem calls that used to surround blitSprite in renderCrosshair are now included in the RenderType passed to it as parameter, so I removed them here.

Thank you for the heads up!

I did some minor style adjustments and fixes. IMO its good to be merged now. Thank you for implementing this feature! Please let me know if you want to test the PR again, prior to merge.

Estecka commented:
Looks good to me!
